lib/python/treadmill_aws/awscredential.py

```diff
@@ -20,59 +23,88 @@
 
-    # no __init__ method.
-    #
-    # pylint: disable=W0232
+    def _parse_name(princ):
+        match = re.search('^([^/@]*)/?([^@]*)@?(.*)$', princ)
+        if match is None:
+            _LOGGER.error((
+                'Internal error - expected name to be a kerberos '
+                'principal name, got [%s]'), princ)
+            raise ValueError('Internal error')
 
-    # pylint: disable=W0613
-    def _error_str(self, string):
-        return str("ERROR: %s" % string).encode("utf-8")
+        name = match.group(1)
+        inst = match.group(2)
+        realm = match.group(3)
+
+        return [name, inst, realm]
 
     class AWSCredentialServer(gssapiprotocol.GSSAPILineServer):
         """AWS Credential server."""
 
-        def __init__(self, accountid, adminprinc):
-            self.accountid = accountid
-            self.adminprinc = adminprinc
+        def __init__(self, account_id, admin_group, realm):
+            self.account_id = account_id
+            self.admin_group = admin_group
+            self.realm = realm
             gssapiprotocol.GSSAPILineServer.__init__(self)
 
         def _validate_request(self, request):
 
-            # should be a valid IAM role (for now)
+            # should verify it is a valid IAM role
             # but it will fail downstream if it is not
-            # should probably verify character set and length
-            pass
+            # for now we just validate that it is a valid user
+            try:
+                uid = pwd.getpwnam(request)
+            except KeyError:
+                raise ValueError("user [%s] is not defined" % request)
 
         def _authorize(self, requestor, request):
 
-            if requestor == self.adminprinc:
+            # admin can request anything (hence request is not considered)
+            if self._authorize_admin(requestor):
                 return
 
-            if '@' not in requestor:
-                _LOGGER.error((
-                    'Internal error - expected requestor to be an kerberos'
-                    'principal name, e.g., user@realm, got [%s]'), requestor)
-                raise ValueError('Internal error')
+            self._authorize_self(requestor, request)
+
+        def _authorize_admin(self, requestor):
+
+            if admin_group is None:
+                return False
+
+            requestor_name, requestor_inst, requestor_realm = \
+                _parse_name(requestor)
+
+            try:
+                if requestor_name in grp.getgrnam(self.admin_group).gr_mem \
+                        and requestor_inst == '' \
+                        and requestor_realm == self.realm:
+                    return True
+            except KeyError as err:
+                _LOGGER.warning(
+                    'admin group [%s] does not exist', self.admin_group)
+
+            return False
+
+        def _authorize_self(self, requestor, request):
+
+            requestor_name, requestor_inst, requestor_realm = \
```

Comment:
fits the line if you rename to

name, instance, realm = _parse_name(requestor)